src/libraries/Microsoft.NETCore.Platforms/src/RID.cs

```diff
@@ -40,20 +47,153 @@ public override string ToString()
+                            if (-1 != runtimeIdentifier.IndexOf(VersionDelimiter, i))
+                            {
+                                break;
+                            }
+                            SetPart();
+                            partStart = i + 1;  // skip delimiter
+                            parseState = RIDPart.Architcture;
+                        }
+                        break;
+                    case RIDPart.Version:
+                        if (current == ArchitectureDelimiter)
+                        {
+                            SetPart();
+                            partStart = i + 1;  // skip delimiter
+                            parseState = RIDPart.Architcture;
+                        }
+                        break;
+                    case RIDPart.Architcture:
+                        if (current == QualifierDelimiter)
+                        {
+                            SetPart();
+                            partStart = i + 1;  // skip delimiter
+                            parseState = RIDPart.Qualifier;
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            partLength = runtimeIdentifier.Length - partStart;
+            if (partLength > 0)
+            {
+                SetPart();
+            }
+
+            string GetPart(RIDPart part)
+            {
+                return parts[(int)part];
+            }
+
+            void SetPart()
+            {
+                if (partLength == 0)
+                {
+                    throw new ArgumentException($"unexpected delimiter at position {partStart} in {runtimeIdentifier}");
+                }
+
+                parts[(int)parseState] = runtimeIdentifier.Substring(partStart, partLength);
+            }
+
+            string version = GetPart(RIDPart.Version);
+
+            if (version == null)
+            {
+                omitVersionDelimiter = false;
+            }
+
+            return new RID()
+            {
+                BaseRID = GetPart(RIDPart.Base),
+                OmitVersionDelimiter = omitVersionDelimiter,
+                Version = version == null ? null : new RuntimeVersion(version),
+                Architecture = GetPart(RIDPart.Architcture),
+                Qualifier = GetPart(RIDPart.Qualifier)
+            };
+        }
+
+        public bool HasVersion => Version != null;
+
+        public bool HasArchitecture => Architecture != null;
+
+        public bool HasQualifier => Qualifier != null;
+
+        public override bool Equals(object obj)
         {
-            return Version != null;
+            return Equals(obj as RID);
         }
 
-        public bool HasArchitecture()
+        public bool Equals(RID obj)
         {
-            return Architecture != null;
+            return object.ReferenceEquals(obj, this) ||
+                (!(obj is null) &&
```

Comment:
can we use `is not null`? 
```suggestion
                (obj is not null &&
```

Reply:
An opportunity for https://github.com/dotnet/runtime/blob/f6b344b64a367ad71df11782122716b2e616d88f/src/libraries/System.Private.CoreLib/src/System/Version.cs#L159 where I copied this from.